### PR TITLE
VEN-143 | Pier mutations

### PR DESCRIPTION
**Pull request by a-rmz**
## Description :sparkles:
* Add mutation and tests for the `Pier` resource.

## Issues :bug:
Partially closes VEN-143

## Testing :alembic:
To manually test the mutations, you can run the following queries on GraphiQL with the inputs provided. Just add valid IDs and you're ready to test. :rocket:

```graphql
query FetchPier($fetchId: ID!) {
  pier(id: $fetchId) {
    id
    properties {
      harbor {
      	id
        properties {
          name
        }
      }
      identifier
      electricity
      gate
      lighting
      mooring
      water
      wasteCollection
      suitableBoatTypes {
        id
        name
      }
    }
  }
}

mutation CreatePier($createInput: CreatePierMutationInput!) {
  createPier(input: $createInput) {
    pier {
      id
    	properties {
        harbor {
          id
        }
        identifier
        electricity
        gate
        lighting
        mooring
        water
        wasteCollection
        suitableBoatTypes {
          id
        }
      }
    }
  }
}

mutation UpdatePier($updateInput: UpdatePierMutationInput!) {
  updatePier(input: $updateInput) {
    pier {
      id
    	properties {
        harbor {
          id
        }
        identifier
        electricity
        gate
        lighting
        mooring
        water
        wasteCollection
        suitableBoatTypes {
          id
        }
      }
    }
  }
}

mutation DeletePier($deleteInput: DeletePierMutationInput!) {
  deletePier(input: $deleteInput) {
    __typename
  }
}
```

```json
{
  "fetchId": "",
  "createInput": {
    "harborId": "",
    "lighting": true,
    "wasteCollection": true,
    "mooring": true,
    "suitableBoatTypes": [ "1" ],
    "identifier": "foobar4",
    "location": {
      "type": "Point",
      "coordinates": [66.6, 99.9]
    },
    "water": true,
    "electricity": true,
    "gate": false
  },
  "updateInput": {
    "id": "",
    "suitableBoatTypes": []
  },
  "deleteInput": {
    "id": ""
  }
}
```

### CODE REVIEW

File: resources/schema.py [hunk 1]
```diff
@@ -452,6 +461,55 @@ def mutate_and_get_payload(cls, root, info, **kwargs):
         return DeleteHarborMutation()
 
 
+class PierInput(AbstractAreaSectionInput):
+    harbor_id = graphene.ID()
+    suitable_boat_types = graphene.List(graphene.String)
```
Comment (igordavydsson) [inline resources/schema.py]: list of IDs?
```python
suitable_boat_types = graphene.List(graphene.ID)
```

File: resources/schema.py [hunk 2]
```diff
@@ -452,6 +461,55 @@ def mutate_and_get_payload(cls, root, info, **kwargs):
...
+            harbor_global_id = from_global_id(harbor_id)[1]
+            try:
+                kwargs["harbor"] = Harbor.objects.get(pk=harbor_global_id)
+            except Harbor.DoesNotExist as e:
+                raise VenepaikkaGraphQLError(e)
+
+        boat_types = set([])
+        for boat_type_id in suitable_boat_types:
+            try:
+                boat_type = BoatType.objects.get(pk=boat_type_id)
```
Comment (igordavydsson) [inline resources/schema.py]: would this work:
```python
suitable_boat_types = kwargs.pop("suitable_boat_types", [])

boat_types = []
try:
    boat_types = BoatType.objects.get(pk__in=suitable_boat_types)
except BoatType.DoesNotExist as e:
    raise VenepaikkaGraphQLError(e)

...
```

File: resources/tests/test_resources_mutations.py [hunk 3]
```diff
@@ -837,3 +837,119 @@ def test_update_harbor_municipality_doesnt_exist(harbor, superuser):
...
+    }
+    assert Pier.objects.count() == 0
+
+    executed = client.execute(
+        query=CREATE_PIER_MUTATION,
+        variables=variables,
+        graphql_url=GRAPHQL_URL,
+        user=superuser,
+    )
+    print(executed)
```
Comment (igordavydsson) [inline resources/tests/test_resources_mutations.py]: 👀

File: resources/schema.py (hunk 1 above)
Comment (a-rmz) [reply resources/schema.py]: The `BoatType` query uses plain IDs, not GQL Global IDs

```graphql
query BT {
  boatTypes {
    id
  }
}
```

```json
{
  "data": {
    "boatTypes": [
      {
        "id": "1"
      },
      {
        "id": "2"
      },
      {
        "id": "3"
      },
      {
        "id": "4"
      },
      {
        "id": "5"
      },
      {
        "id": "6"
      },
      {
        "id": "7"
      },
      {
        "id": "8"
      }
    ]
  }
}
```

File: resources/schema.py (hunk 2 above)
Comment (a-rmz) [reply resources/schema.py]: Yeah, would work if there was a way to figure out if there was an invalid one.

File: resources/schema.py (hunk 1 above)
Comment (igordavydsson) [reply resources/schema.py]: yes, plain ID can be a string or an integer in graphene

Comment (igordavydsson) [reply resources/schema.py]: I think this might be valid list `["1", 2, "3"]` 🤔

File: resources/schema.py [hunk 4]
```diff
@@ -452,6 +461,55 @@ def mutate_and_get_payload(cls, root, info, **kwargs):
...
+
+        harbor_id = kwargs.pop("harbor_id", None)
+        if harbor_id:
+            harbor_global_id = from_global_id(harbor_id)[1]
+            try:
+                kwargs["harbor"] = Harbor.objects.get(pk=harbor_global_id)
+            except Harbor.DoesNotExist as e:
+                raise VenepaikkaGraphQLError(e)
+
+        boat_types = set([])
```
Comment (igordavydsson) [inline resources/schema.py]: I think if you just want to initiate an empty set, you could just do `set()`

File: resources/schema.py (hunk 2 above)
Comment (igordavydsson) [reply resources/schema.py]: yeah, ok, let's keep it like this

File: resources/tests/test_resources_mutations.py [hunk 5]
```diff
@@ -837,3 +837,118 @@ def test_update_harbor_municipality_doesnt_exist(harbor, superuser):
...
     assert Harbor.objects.count() == 1
     assert_doesnt_exist("Municipality", executed)
+
+
+CREATE_PIER_MUTATION = """
+mutation CreatePier($input: CreatePierMutationInput!) {
+  createPier(input: $input) {
+    pier {
+      id
+        properties {
```
Comment (igordavydsson) [inline resources/tests/test_resources_mutations.py]: indentation? 👀

File: resources/schema.py [hunk 6]
```diff
@@ -510,6 +510,49 @@ def mutate_and_get_payload(cls, root, info, **kwargs):
...
+            try:
+                boat_type = BoatType.objects.get(pk=boat_type_id)
+            except BoatType.DoesNotExist as e:
+                raise VenepaikkaGraphQLError(e)
+            boat_types.add(boat_type)
+
+        try:
+            update_object(pier, kwargs)
+            pier.suitable_boat_types.set(boat_types)
+            pier.save()
```
Comment (igordavydsson) [inline resources/schema.py]: I think `save()` is not needed after changing m2m relations, as they are saved immediately.

File: resources/schema.py [hunk 7]
```diff
@@ -510,6 +510,49 @@ def mutate_and_get_payload(cls, root, info, **kwargs):
...
+    @transaction.atomic
+    def mutate_and_get_payload(cls, root, info, **kwargs):
+        # TODO: Should check if the user has permissions to
+        # delete the specific resource
+        id = from_global_id(kwargs.pop("id"))[1]
+
+        try:
+            harbor_id = kwargs.pop("harbor_id", None)
+            if harbor_id:
+                global_harbor_id = from_global_id(harbor_id)[1]
```
Comment (igordavydsson) [inline resources/schema.py]: well, this is actually `harbor_id`, but the var above could be called `global_harbor_id` 🙂

File: resources/schema.py (hunk 2 above)
Comment (a-rmz) [reply resources/schema.py]: Another option would be to compare something like this:
```python
set(suitable_boat_types).difference(
    set(Model.objects.filter(id__in=suitable_boat_types).values_list('id'),
    flat=True)
)
```
And if there's a difference, we can raise the error saying that values X, Y, &Z  are missing.

Comment (igordavydsson) [reply resources/schema.py]: oh, pretty neat idea! probably a bit of an overengineering for this case, and this is still a hit to the db, but slightly less readable that your current solution. let's keep things as they are 🙂

File: resources/schema.py (hunk 1 above)
Comment (a-rmz) [reply resources/schema.py]: Just FTR, it is a [valid id](https://docs.graphene-python.org/en/latest/types/scalars/).
```
graphene.ID represents a unique identifier, often used to refetch an
object or as key for a cache. The ID type appears in a JSON
response as a String; however, it is not intended to be
human-readable. When expected as an input type, any string
(such as “4”) or integer (such as 4) input value will be accepted as an ID.
```

Comment (igordavydsson) [reply resources/schema.py]: yeah, that's what I meant. 🙂

**Pull request merged**